### PR TITLE
Sharey keyword for boxplot

### LINKED ISSUE #20918: Please, add sharey keyword to boxplot.
#### Code Sample, a copy-pastable example if possible

```python
df.groupby('some feature').boxplot(sharey=False)
```
#### Problem description

Boxplot by default creates plots with shared y axis. The keyword `sharey=False` is not accepted. However, it would be consistent with df.plot(). This is unfortunate if you don't want to compare values across the subplots. There is a workaround mentioned [here](https://stackoverflow.com/questions/40148830/different-ylim-for-shared-axes-in-pandas-boxplot), but having the option to turn that "feature" off would be much simpler.  

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-17134-Microsoft
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: None
pip: 9.0.3
setuptools: 39.0.1
Cython: None
numpy: 1.14.2
scipy: 1.0.1
pyarrow: None
xarray: None
IPython: 6.3.1
sphinx: None
patsy: 0.5.0
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: None
tables: 3.4.3
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: 1.2.7
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

​
</details>

### PR DESCRIPTION
**Pull request by sorenwacker**
…to _subplots().

.. code-block:: jupyter-notebook

   %pylab inline
    import pandas as pd

    N = 100
    rand = random.random(N)
    clas = random.binomial(5,.5, N)
    df = pd.DataFrame({'Rand': rand-clas,
                    'Rand2': rand,
                    'Class': clas},
                    index= np.arange(N))

    df.groupby('Class').boxplot(sharey=True, sharex=False)
    >>> TypeError: boxplot() got an unexpected keyword argument 'sharey'

New Behavior:

.. ipython:: jpyter-notebook:

    ...

    df.groupby('Class').boxplot(sharey=True, sharex=True)
    df.groupby('Class').boxplot(sharey=True, sharex=False)
    df.groupby('Class').boxplot(sharey=False, sharex=True)
    df.groupby('Class').boxplot(sharey=False, sharex=False)

    All leads to different behaviour. The shareing of axes both x and y
    can be turned on and off separately. 

To restore previous behavior, use boxplot() without keywords.

Default is the previous behavior of sharey
- [x ] closes #20918
- [x ] tests added / passed
- [ x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x ] whatsnew entry

### CODE REVIEW

**Comment by TomAugspurger** [comment]
We'll want tests for this. See the e.g. `test_sharex_and_ax` in pandas/tests/plotting/test_frame.py for an example.

**Comment by soerendip** [comment]
Thanks. I will look into that and give it a try.

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -898,6 +898,48 @@ yourself. To revert to the old setting, you can run this line:
 
   pd.options.display.max_columns = 20
 
+.. _whatsnew_0230.boxplot.sharexy:
+
+Optional sharing of x/y-axis by pandas.DataFrame().groupby().boxplot()
```
[the remark this answers is not in the document]
Comment (soerendip) [reply doc/source/whatsnew/v0.23.0.txt]: I updated the docstring and the the part in v0.23.0.txt.

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply doc/source/whatsnew/v0.23.0.txt]: Can you move this to 0.23.1.txt now?

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply doc/source/whatsnew/v0.23.0.txt]: And I think the example can be trimmed down. I'd rather update the docstring for `DataFrame.plot.box` since that's a longer-term thing. The release note are for people checking changes. They'll see that it now supports `sharey`, and click through if interested.

File: doc/source/whatsnew/v0.23.0.txt [hunk 2]
```diff
@@ -898,6 +898,48 @@ yourself. To revert to the old setting, you can run this line:
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+(:issue:`15184`)
+
+Previous Behavior:
+
+.. code-block:: jupyter-notebook
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: See the other examples in `/pandas/plotting/_core.py` for how these should be formatted.

File: doc/source/whatsnew/v0.23.0.txt [hunk 3]
```diff
@@ -898,6 +898,48 @@ yourself. To revert to the old setting, you can run this line:
...
+
+    Using boxplot with keywords sharex or sharey resulted in an error.
+
+New Behavior:
+
+.. ipython:: jpyter-notebook:
+    
+    ...
+
+    df.groupby('Class').boxplot(sharey=True, sharex=True)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: Won't need all of these. Probably just one will convey the point.

Comment (soerendip) [reply doc/source/whatsnew/v0.23.0.txt]: Ok,  0.23.1.txt does not exist yet. When I fetch from upstream I don't receive it. Does that mean I should create that file? I thought I already removed the example. Not sure why it still there.

**Comment by soerendip** [comment]
Done. Hope it is correct this time.

File: doc/source/whatsnew/v0.23.0.txt (hunk 3 above)
Comment (WillAyd) [reply doc/source/whatsnew/v0.23.0.txt]: @soerendip did you merge after fetching? The file is there already on master

File: pandas/plotting/_core.py [hunk 4]
```diff
@@ -2567,6 +2567,12 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
     figsize : A tuple (width, height) in inches
     layout : tuple (optional)
         (rows, columns) for the layout of the plot
+    sharex :
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py]: Could you take a look at http://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html for how to format these? Something like

```
sharex : bool, default False
    Whether the horizontal axes is shared among subplots
```

File: pandas/tests/plotting/test_frame.py [hunk 5]
```diff
@@ -367,6 +367,64 @@ def test_subplots(self):
...
+        # expected them, so make sure that only the right ones are removed
+
+        df = DataFrame({'a': [-1.43, -0.15, -3.70, -1.43, -0.14],
+                        'b': [0.56, 0.84, 0.29, 0.56, 0.85],
+                        'c': [0, 1, 2, 3, 1]},
+                       index=[0, 1, 2, 3, 4])
+
+        # standart behavior
+        axes = df.groupby('c').boxplot()
+        self._check_visible(axes[0].get_yticklabels(), visible=True)
```
Comment (TomAugspurger) [inline pandas/tests/plotting/test_frame.py]: These tests are a bit wordy. I wonder if the following is clear?

```python
expected = pd.Series([True, False, True, False])
result = axes.apply(lambda ax: ax.yaxis.get_visible())
tm.assert_frame_equal(result, expected)
```

Is that the same? Do you find it clearer?

Comment (soerendip) [reply pandas/tests/plotting/test_frame.py]: I see where you want to go. Would this be acceptable? It uses a function that you might want somewhere else though in case you want to use it in other tests? Where would be a good position to put it?

    def test_groupby_boxplot_sharex(self):

        def _assert_xtickslabels_visibility(axes, expected):
            for ax, exp in zip(axes, expected):
                self._check_visible(ax.get_xticklabels(), visible=exp)
    
        df = DataFrame({'a': [-1.43, -0.15, -3.70, -1.43, -0.14],
                        'b': [0.56, 0.84, 0.29, 0.56, 0.85],
                        'c': [0, 1, 2, 3, 1]},
                       index=[0, 1, 2, 3, 4])

        # standart behavior
        axes = df.groupby('c').boxplot()
        expected = [True, True, True, True]
        _assert_xtickslabels_visibility(axes, expected)
        # set sharex=False should be identical
        axes = df.groupby('c').boxplot(sharex=False)
        expected = [True, True, True, True]
        _assert_xtickslabels_visibility(axes, expected)        
        # sharex=True, xticklabels should be visible only for bottom plots
        axes = df.groupby('c').boxplot(sharex=True)
        expected = [False, False, True, True]
        _assert_xtickslabels_visibility(axes, expected)´

File: pandas/plotting/_core.py (hunk 4 above)
Comment (soerendip) [reply pandas/plotting/_core.py]: Ok, did that. Should this part of the doc-string then be corrected as well?

    subplots :
        * ``False`` - no subplots will be used
        * ``True`` - create a subplot for each group

**Comment by soerendip** [comment]
Any updates on this?

File: doc/source/whatsnew/v0.23.0.txt [hunk 6]
```diff
@@ -927,6 +927,14 @@ yourself. To revert to the old setting, you can run this line:
 
   pd.options.display.max_columns = 20
 
+.. _whatsnew_0230.boxplot.sharexy:
+
+Optional sharing of x/y-axis by pandas.DataFrame().groupby().boxplot()
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+(:issue:`15184`)
+
+Optional sharing of x/y-axis by pandas.DataFrame().groupby().boxplot()
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: needs to move to 0.24.0

File: doc/source/whatsnew/v0.23.1.txt [hunk 7]
```diff
@@ -76,7 +76,7 @@ I/O
 Plotting
 ^^^^^^^^
 
--
+- Optional sharing of x/y-axis by pandas.DataFrame().groupby().boxplot() (:issue: `20968`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: remove from here

File: pandas/plotting/_core.py [hunk 8]
```diff
@@ -2567,6 +2567,10 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
     figsize : A tuple (width, height) in inches
     layout : tuple (optional)
         (rows, columns) for the layout of the plot
+    sharex : bool, default False
+        Whether x-axes will be shared among subplots
```
Comment (jreback) [inline pandas/plotting/_core.py]: add a versionadded tag (0.24.0)

File: pandas/tests/plotting/test_frame.py [hunk 9]
```diff
@@ -367,6 +367,61 @@ def test_subplots(self):
             for ax in axes:
                 assert ax.get_legend() is None
 
+    def test_groupby_boxplot_sharey(self):
+        # https://github.com/pandas-dev/pandas/issues/20968
+        # sharey can now be switched check whether the right
+        # pair of axes is turned on or off
+
+        def _assert_ytickslabels_visibility(axes, expected):
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: do we have similar functions at a top level? maybe we should

File: pandas/tests/plotting/test_frame.py [hunk 10]
```diff
@@ -367,6 +367,61 @@ def test_subplots(self):
...
+
+        df = DataFrame({'a': [-1.43, -0.15, -3.70, -1.43, -0.14],
+                        'b': [0.56, 0.84, 0.29, 0.56, 0.85],
+                        'c': [0, 1, 2, 3, 1]},
+                       index=[0, 1, 2, 3, 4])
+
+        # standart behavior
+        axes = df.groupby('c').boxplot()
+        expected = [True, False, True, False]
+        _assert_ytickslabels_visibility(axes, expected)
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: use a blank line between cases

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/plotting/test_frame.py (hunk 9 above)
Comment (soerendip) [reply pandas/tests/plotting/test_frame.py]: Sorry, I am new to pandas. I am not sure whether I understand what you mean. Are you saying, I should move the function somewhere else? If yes, where exactly should the function go?

File: pandas/plotting/_core.py (hunk 8 above)
Comment (TomAugspurger) [reply pandas/plotting/_core.py]:
```
    sharex : bool, default False
        Whether x-axes will be shared among subplots 

        .. versionadded:: 0.23.1

**Comment by soerendip** [comment]
All changes should be done.

File: doc/source/whatsnew/v0.23.0.txt [hunk 11]
```diff
@@ -927,6 +927,14 @@ yourself. To revert to the old setting, you can run this line:
 
   pd.options.display.max_columns = 20
 
+.. _whatsnew_0230.boxplot.sharexy:
+
+Optional sharing of x/y-axis by pandas.DataFrame().groupby().boxplot()
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+(:issue:`15184`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: this file needs to be reverted

File: pandas/tests/plotting/test_frame.py (hunk 9 above)
Comment (jreback) [reply pandas/tests/plotting/test_frame.py]: you can make this a function on the class itself rather than defininig it inside the function

File: pandas/tests/plotting/test_frame.py [hunk 12]
```diff
@@ -367,6 +367,65 @@ def test_subplots(self):
...
+        def _assert_ytickslabels_visibility(axes, expected):
+            for ax, exp in zip(axes, expected):
+                self._check_visible(ax.get_yticklabels(), visible=exp)
+
+        df = DataFrame({'a': [-1.43, -0.15, -3.70, -1.43, -0.14],
+                        'b': [0.56, 0.84, 0.29, 0.56, 0.85],
+                        'c': [0, 1, 2, 3, 1]},
+                       index=[0, 1, 2, 3, 4])
+
+        # standart behavior
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: standard

Comment (jreback) [reply pandas/tests/plotting/test_frame.py]: what is 'standard' mean here, can you update the comment

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/plotting/test_frame.py (hunk 12 above)
Comment (soerendip) [reply pandas/tests/plotting/test_frame.py]: Done.

**Comment by TomAugspurger** [comment]
@soerendip looks like tests are failing in the latest commit.

**Comment by soerendip** [comment]
That’s when you think “it’s so easy it can’t fail”... will look into this.

________________________________

File: doc/source/whatsnew/v0.23.1.txt (hunk 7 above)
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: its not clear to me what is being fixed / enhancement here, can you re-word

File: pandas/tests/plotting/test_frame.py [hunk 13]
```diff
@@ -2921,6 +2972,14 @@ def test_secondary_axis_font_size(self, method):
         self._check_ticks_props(axes=ax.right_ax,
                                 ylabelsize=fontsize)
 
+    def _assert_ytickslabels_visibility(self, axes, expected):
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: you you move these to the very top of the class (I think we hvae other checker functions there)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by soerendip** [comment]
Done.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @soerendip
